File: generated-code/cpp/codecraft/TcpReadWrite/model/RepairProperties.cpp
#include "RepairProperties.hpp"

namespace model {

RepairProperties::RepairProperties(std::vector<model::EntityType> validTargets, int power) : validTargets(validTargets), power(power) { }

// Read RepairProperties from input stream
RepairProperties RepairProperties::readFrom(InputStream& stream) {
    std::vector<model::EntityType> validTargets = std::vector<model::EntityType>();
    size_t validTargetsSize = stream.readInt();
    validTargets.reserve(validTargetsSize);
    for (size_t validTargetsIndex = 0; validTargetsIndex < validTargetsSize; validTargetsIndex++) {
        model::EntityType validTargetsElement = readEntityType(stream);
        validTargets.emplace_back(validTargetsElement);
    }
    int power = stream.readInt();
    return RepairProperties(validTargets, power);
}

// Write RepairProperties to output stream
void RepairProperties::writeTo(OutputStream& stream) const {
    stream.write((int)(validTargets.size()));
    for (const model::EntityType& validTargetsElement : validTargets) {
        stream.write((int)(validTargetsElement));
    }
    stream.write(power);
}

// Get string representation of RepairProperties
std::string RepairProperties::toString() const {
    std::stringstream ss;
    ss << "RepairProperties { ";
    ss << "validTargets: ";
    ss << "[ ";
    for (size_t validTargetsIndex = 0; validTargetsIndex < validTargets.size(); validTargetsIndex++) {
        const model::EntityType& validTargetsElement = validTargets[validTargetsIndex];
        if (validTargetsIndex != 0) {
            ss << ", ";
        }
        ss << entityTypeToString(validTargetsElement);
    }
    ss << " ]";
    ss << ", ";
    ss << "power: ";
    ss << power;
    ss << " }";
    return ss.str();
}

}

File: generated-code/ruby/codecraft/FileReadWrite/model/repair_properties.rb
require './model/entity_type'

module Model

# Entity's repair properties
class RepairProperties
    # Valid target entity types
    attr_accessor :valid_targets
    # Health restored in one tick
    attr_accessor :power

    def initialize(valid_targets, power)
        @valid_targets = valid_targets
        @power = power
    end

    # Read RepairProperties from input stream
    def self.read_from(stream)
        valid_targets = []
        stream.read_int().times do |_|
            valid_targets_element = Model::EntityType.read_from(stream)
            valid_targets.push(valid_targets_element)
        end
        power = stream.read_int()
        RepairProperties.new(valid_targets, power)
    end

    # Write RepairProperties to output stream
    def write_to(stream)
        stream.write_int(@valid_targets.length())
        @valid_targets.each do |valid_targets_element|
            stream.write_int(valid_targets_element)
        end
        stream.write_int(@power)
    end

    def to_s
        string_result = "RepairProperties { "
        string_result += "valid_targets: "
        string_result += "[ "
        valid_targets_index = 0
        @valid_targets.each do |valid_targets_element|
            if valid_targets_index != 0
                string_result += ", "
            end
            string_result += EntityType.to_s(valid_targets_element)
            valid_targets_index += 1
        end
        string_result += " ]"
        string_result += ", "
        string_result += "power: "
        string_result += @power.to_s
        string_result += " }"
        string_result
    end

    def to_str
        to_s
    end
end

end